packages/nbdime/src/common/mergeview.ts

```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
     mergeview.base.model.mimeType = mimetype;
   }
   return mergeview;
 }
 
-
 /**
  * Used by MergeView to show diff in a string diff model
  */
-export
-class DiffView {
-  constructor(model: IStringDiffModel,
-              type: 'left' | 'right' | 'merge',
-              updateCallback: (force?: boolean) => void,
-              options: CodeMirror.MergeView.MergeViewEditorConfiguration) {
+export class DiffView {
+  constructor(
+    model: IStringDiffModel,
+    type: 'left' | 'right' | 'merge',
+    listener: Extension,
+    mergeControlGutter: Extension,
+    options: IMergeViewEditorConfiguration
+  ) {
     this.model = model;
     this.type = type;
-    this.updateCallback = updateCallback;
-    this.classes = type === 'left' ?
-      leftClasses : type === 'right' ? rightClasses : null;
-    let ownValue = this.model.remote || '';
-    this.ownWidget = new EditorWidget(ownValue, copyObj({readOnly: !!options.readOnly}, options));
+    let remoteValue = this.model.remote || '';
+    this._remoteEditorWidget = new EditorWidget(remoteValue); // OPTIONS TO BE GIVEN
+    this._remoteEditorWidget.editor.injectExtension([listener, mergeControlGutter, getCommonEditorExtensions()]);
     this.showDifferences = options.showDifferences !== false;
   }
 
-  init(base: CodeMirror.Editor) {
-    this.baseEditor = base;
-    (this.baseEditor.state.diffViews ||
-     (this.baseEditor.state.diffViews = [])).push(this);
-    this.ownEditor.state.diffViews = [this];
-
-    this.baseEditor.on('gutterClick', this.onGutterClick.bind(this));
-    this.ownEditor.on('gutterClick', this.onGutterClick.bind(this));
-
+  init(baseWidget: EditorWidget) {
+    this.baseEditorWidget = baseWidget;
+    let baseEditor = this.baseEditorWidget.cm;
+    let remoteEditor = this.remoteEditorWidget.cm;
     this.lineChunks = this.model.getLineChunks();
     this.chunks = lineToNormalChunks(this.lineChunks);
-    this.dealigned = false;
-
-    this.forceUpdate = this.registerUpdate();
-    this.setScrollLock(true, false);
-    this.registerScroll();
+    this.updateView(baseEditor, remoteEditor);
   }
 
-  setShowDifferences(val: boolean) {
-    val = val !== false;
-    if (val !== this.showDifferences) {
-      this.showDifferences = val;
-      this.forceUpdate('full');
-    }
+  updateView( baseEditor: EditorView, remoteEditor: EditorView) {
+    this.clearHighlighting(
+      remoteEditor,
+      this.model.additions,
+      this.chunks,
+      DIFF_OP.DIFF_INSERT
+    );
+
+    this.clearHighlighting(
+      baseEditor,
+      this.model.deletions,
+      this.chunks,
+      DIFF_OP.DIFF_DELETE
+    );
+
+    this.updateHighlighting(
+      remoteEditor,
+      this.model.additions,
+      this.chunks,
+      DIFF_OP.DIFF_INSERT
+    );
+
+    this.updateHighlighting(
+      baseEditor,
+      this.model.deletions,
+      this.chunks,
+      DIFF_OP.DIFF_DELETE
+    );
+
+    this.syncScroll(baseEditor, remoteEditor);
+    this.syncScroll(remoteEditor, baseEditor);
```

Comment:
I think these functions should be called in init instead, no need to add a new listener each time the view updates.

Reply:
Thanks. This has been changed.